include/dlaf/communication/communicator_pipeline.h

```diff
@@ -0,0 +1,141 @@
+//
+
+/// @file
+
+#pragma once
+
+#include <dlaf/common/index2d.h>
+#include <dlaf/common/pipeline.h>
+#include <dlaf/common/round_robin.h>
+#include <dlaf/communication/communicator.h>
+#include <dlaf/communication/communicator_type.h>
+#include <dlaf/communication/index.h>
+
+namespace dlaf::comm {
+namespace internal {
+constexpr const dlaf::common::Ordering FULL_COMMUNICATOR_ORDER{dlaf::common::Ordering::RowMajor};
+}
+
+using CommunicatorPipelineSharedWrapper = typename common::Pipeline<Communicator>::ReadOnlyWrapper;
+using CommunicatorPipelineExclusiveWrapper = typename common::Pipeline<Communicator>::ReadWriteWrapper;
+using CommunicatorPipelineSharedSender = typename common::Pipeline<Communicator>::ReadOnlySender;
+using CommunicatorPipelineExclusiveSender = typename common::Pipeline<Communicator>::ReadWriteSender;
+
+/// A CommunicatorPipeline provides pipelined access to a communicator, as well as metadata about the
+/// communicator.
+///
+/// A CommunicatorPipeline uses a Pipeline to provide shared or exclusive access through senders to the
+/// underlying communicator.  @tparam CT signals whether the contained communicator is a row, column, or
+/// full grid communicator.
+template <CommunicatorType CT>
+class CommunicatorPipeline {
+  using PipelineType = dlaf::common::Pipeline<Communicator>;
+
+public:
+  using SharedWrapper = typename PipelineType::ReadOnlyWrapper;
+  using ExclusiveWrapper = typename PipelineType::ReadWriteWrapper;
+  using SharedSender = typename PipelineType::ReadOnlySender;
+  using ExclusiveSender = typename PipelineType::ReadWriteSender;
+
+  static constexpr CommunicatorType communicator_type = CT;
+
+  /// Create a CommunicatorPipeline by moving in the resource (it takes the
+  /// ownership).
+  explicit CommunicatorPipeline(Communicator comm, Index2D rank = {-1, -1}, Size2D size = {-1, -1})
+      : pipeline_(std::move(comm)), rank_(std::move(rank)), size_(std::move(size)) {}
+  CommunicatorPipeline(CommunicatorPipeline&& other) = default;
+  CommunicatorPipeline& operator=(CommunicatorPipeline&& other) = default;
+  CommunicatorPipeline(const CommunicatorPipeline&) = delete;
+  CommunicatorPipeline& operator=(const CommunicatorPipeline&) = delete;
+
+  /// Return the rank of the current process in the CommunicatorPipeline.
+  IndexT_MPI rank() const noexcept {
+    return rankFullCommunicator(rank_);
+  }
+
+  /// Return the size of the grid.
+  IndexT_MPI size() const noexcept {
+    return size_.linear_size();
+  }
+
+  /// Return the 2D rank of the current process in the grid that this pipeline
+  /// belongs to.
+  Index2D rank_2d() const noexcept {
+    return rank_;
+  }
+
+  /// Return the 2D size of the current process in the grid that this pipeline
+  /// belongs to.
+  Size2D size_2d() const noexcept {
+    return size_;
+  }
+
+  /// Return rank in the grid with all ranks given the 2D index.
+  IndexT_MPI rankFullCommunicator(const Index2D& index) const noexcept {
```

Comment:
why not snake_case?

Reply:
Because of copy-paste. This comes from `CommunicatorGrid`. Good catch!

Reply:
Updated in https://github.com/eth-cscs/DLA-Future/pull/993/commits/4927179823baf6ffa10357482e47390fab145ddf. Note that I'm not changing the camelCase members in `CommunicatorGrid` here.